overlord/servicestate/quota_handlers.go

```diff
@@ -823,7 +854,75 @@ func validateSnapForAddingToGroup(st *state.State, snaps []string, group string,
 			}
 		}
 	}
+	return nil
+}
 
+// splitSnapServiceName splits and verifies the snap service reference
+// taken in by the frontend. It expects the format snap.service
+func splitSnapServiceName(name string) (string, string, error) {
+	parts := strings.Split(name, ".")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid snap service: %s", name)
+	}
+	if err := naming.ValidateSnap(parts[0]); err != nil {
+		return "", "", err
+	}
+	return parts[0], parts[1], naming.ValidateApp(parts[1])
+}
+
+// appReferenceIsService returns whether the service referred to in the
+// snap is actually a service.
+func appReferenceIsService(st *state.State, snap, service string) error {
+	snapInfo, err := snapstate.CurrentInfo(st, snap)
+	if err != nil {
+		return err
+	}
+
+	appInfo, ok := snapInfo.Apps[service]
+	if !ok {
+		return fmt.Errorf("invalid service %q", service)
+	}
+	if !appInfo.IsService() {
+		return fmt.Errorf("%q is not a service", service)
+	}
+	return nil
+}
+
+// ensureGroupHasNoSubgroups returns true if the group has no sub-groups or it doesn't exist,
+// otherwise turns false.
+func ensureGroupHasNoSubgroups(group string, allGrps map[string]*quota.Group) bool {
+	grp, ok := allGrps[group]
+	if ok && len(grp.SubGroups) != 0 {
+		return false
+	}
+	return true
+}
+
+func validateSnapServicesForAddingToGroup(st *state.State, services []string, group string, parentGroup *quota.Group, allGrps map[string]*quota.Group) error {
+	// Services come in the format of snap.service, and we do the split and parsing
+	// in this function. We need to make sure that the snaps referred are related to
+	// the current group, and we make sure that the services are valid service names.
```

Comment:
I would move this comment on top of the function. Also adding a line saying that the state needs to be locked might be a good idea.

Reply:
Thank you! Will correct this!